File: Graph/python/02_우주신과의교감.py
# fast campus 강의
# https://www.acmicpc.net/problem/1774
# 2

import math

def get_distance(p1, p2):
    x = p1[0] - p2[0]
    y = p1[1] - p2[1]
    return math.sqrt((x * x) + (y * y))

def find_parent(x):
    if parent[x] != x:
        parent[x] = find_parent(parent[x])
    return parent[x]

def union_parent(a, b):
    a = find_parent(a)
    b = find_parent(b)
    if a < b:
        parent[a] = b
    else:
        parent[b] = a

N, M = map(int, input().split())
parent = [0] * (N + 1)
locations = []
edges = []

for _ in range(N):
    x, y = map(int, input().split())
    locations.append((x, y))

length = len(locations)
for i in range(length - 1):
    for j in range(i + 1, length):
        edges.append((get_distance(locations[i], locations[j]), i + 1, j + 1))
edges.sort()

for i in range(1, N + 1):
    parent[i] = i

for _ in range(M):
    a, b = map(int, input().split())
    union_parent(a, b)

result = 0
for cost, a, b in edges:
    if find_parent(a) != find_parent(b):
        union_parent(a, b)
        result += cost
print("%0.2f" % result)

File: Graph/c++/02_우주신과의교감.cpp
// fast campus 강의
// https://www.acmicpc.net/problem/1774
// 2

#include <iostream>
#include <algorithm>
#include <vector>
#include <cmath>

using namespace std;

int N, M;
int parent[1001];
vector<pair<int, int>> locations;
vector<pair<double, pair<int, int>>> edges;

double get_distance(pair<int, int> p1, pair<int, int> p2) {
	long long x = p1.first - p2.first;
	long long y = p1.second - p2.second;
	return sqrt((x * x) + (y * y));
}

int find_parent(int a) {
	if (parent[a] == a)
		return a;
	else
		return parent[a] = find_parent(parent[a]);
}

void union_parent(int a, int b) {
	a = find_parent(a);
	b = find_parent(b);
	if (a < b)
		parent[b] = a;
	else
		parent[a] = b;
}

int main() {
	cin >> N >> M;
	for (int i = 0; i < N; i++) {
		int x, y;
		cin >> x >> y;
		locations.push_back({ x, y });
	}

	for (int i = 0; i < N - 1; i++) {
		for (int j = i + 1; j < N; j++)
			edges.push_back({ get_distance(locations[i], locations[j]), {i + 1, j + 1} });
	}
	sort(edges.begin(), edges.end());

	for (int i = 1; i < N + 1; i++)
		parent[i] = i;

	for (int i = 0; i < M; i++) {
		int a, b;
		cin >> a >> b;
		union_parent(a, b);
	}

	double result = 0;
	for (int i = 0; i < edges.size(); i++) {
		double cost = edges[i].first;
		int a = edges[i].second.first;
		int b = edges[i].second.second;
		if (find_parent(a) != find_parent(b)) {
			union_parent(a, b);
			result += cost;
		}
	}

	printf("%0.2f", result);
}